### DISCUSSION TITLE
Solution for cronjobs running out of control

### DISCUSSION THREAD

**Discussion by KayJay89** [dokku/dokku#4818] [Q&A]
Hi all,

I've been using Dokku successfully for a while now. Over the last few months I've been working on one for a few months which runs a cronjob every minute during certain timeframes (4h per weekday).

The script checks 200-300 RSS urls for new items and if new item creates a new entry in my Postgresql db. If all goes well it takes 20-40sec for the script to complete but sometimes it takes 60-70sec. This creates a snowball effect and with every new cronjob launched the CPU recources decline making every job taking much longer.

I would like to know your guys' take on how to approach this best. I have done some reserach and see several options which I'll lay out below. Keep in mind for the moment I am using the tiniest Linode VPS (1GB mem + CPU 1Ghz). This is more than enough during "low tide" when the script runs every 5-10min, I barely reach 10%. Just during those couple hours every day, it gets out of hand. Eventually I'll be running multiple of these jobs simultaneously, so I will definitely upgrade to a larger Linode eventually.

This is the first time I use cron, and I am thus very unfamiliar with cron and suggestions like celery, rabbitmq,..

My "solutions":

**1.** Use a lockfile and create a bash script that first checks if the job is already running prior to firing off another one, and if it is, don't do anything (this seems to be the easiest but also least extensible version)
**2.** Resort to job queues like celery however I read celery is not entirely up to speed anymore. Then I also see suggestions (as in the Dokku docs) to use RabbitMq. But when digging into that on Stackoverflow, some say it's not a scheduler as such.

I'd like to have your input on how I _could_ proceed, not how I _should_ ,with the above in mind. I'd like to see if there is any option I have missed prior to making my decision.

**Comment by KayJay89** [comment]
I just realize that because of the way Dokku creates a new docker container for each cron task, the lock file option won't work.

**Comment by josegonzalez** [comment]
A good alternative might just be to have a persistent process. Your process can just run the command on a timer every 60 seconds. Based on your mentioning of celery, you could use something like threading or multiprocessing to spawn that code every 60 seconds.

RabbitMQ is a good alternative if you want a larger messaging system. It doesn't schedule anything as much as allow for scheduling-type patterns. As far as I know, celery is still a reasonable way to do messaging, and supports a variety of backends.

Finally, while we don't have any such locking code, having a way to prevent overlap of cron tasks could be useful. I would definitely file a ticket for that and we could investigate how such a thing might work.
